### PR TITLE
Euler connector

### LINKED ISSUE #237: Euler Connector
Create Euler Connector with the following functions -
1. deposit
2. withdraw
3. mint
4. burn
5. approve
6. transfer
7. mint
8. burn
9. enter market
10. exit market

### PR DESCRIPTION
**Pull request by shriyatyagii**
**Euler Connector**
Closes #237

Functions Added:
- deposit
- withdraw
- borrow
- repay
- eTransfer
- dTranfer
- mint
- burn
- approveDebt
- swap
- enterMarket
- exitMarket

**Testcases-**
image

### CODE REVIEW

File: contracts/mainnet/connectors/euler/main.sol [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+import { TokenInterface } from "../../common/interfaces.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+abstract contract Euler is Helpers {
+	using SafeERC20 for IERC20;
+
+	/**
+	 * @dev Deposit ETH/ERC20_Token.
+	 * @notice Deposit a token to Euler for lending / collaterization.
+	 * @param subAccount Subaccount number
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
	 * @param subAccount Sub-account Id (0 for primary and 1 - 255 for sub-account)
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+			_amt,
+			enableCollateral,
+			getId,
+			setId
+		);
+	}
+
+	/**
+	 * @dev Withdraw ETH/ERC20_Token.
+	 * @notice Withdraw deposited token Euler
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
	 * @notice Withdraw deposited token and earned interest from Euler
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+
+		IEulerDToken borrowedDToken = IEulerDToken(
+			markets.underlyingToDToken(_token)
+		);
+		borrowedDToken.borrow(subAccount, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		borrowedDToken.borrow(subAccount, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+
+		address _token = isEth ? wethAddr : token;
+		IEulerDToken borrowedDToken = IEulerDToken(
+			markets.underlyingToDToken(_token)
+		);
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: `_amt` can be `uint256(-1)`
```suggestion
		if (isEth) {
			_amt = _amt == uint256(-1) ? borrowedDToken.balanceOf(address(this)) : _amt;
			convertEthToWeth(isEth, TokenInterface(_token), _amt);
		}
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+
+		_amt = _amt == type(uint256).max
+			? borrowedDToken.balanceOf(address(this))
+			: _amt;
+
+		TokenInterface(_token).approve(EULER_MAINNET, _amt);
+		borrowedDToken.repay(subAccount, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		borrowedDToken.repay(subAccount, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+		address _token = isEth ? wethAddr : token;
+		IEulerDToken borrowedDToken = IEulerDToken(
+			markets.underlyingToDToken(_token)
+		);
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		_amt = _amt == type(uint256).max
+			? borrowedDToken.balanceOf(address(this))
+			: _amt;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: We can do this before `isEth`

File: contracts/mainnet/connectors/euler/main.sol [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+
+		TokenInterface(_token).approve(EULER_MAINNET, _amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: better to use the `approve` function.

File: contracts/mainnet/connectors/euler/main.sol [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+
+		setUint(setId, _amt);
+
+		_eventName = "LogRepay(uint256,address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(subAccount, token, _amt, getId, setId);
+	}
+
+	/**
+	 * @dev Mint ETH/ERC20_Token.
+	 * @notice Mint a token from Euler. Maint creates equal amount of deposits and debts.
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
	 * @notice Mint a token from Euler. Mint creates an equal amount of deposits and debts. (self-borrow)
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		eToken.mint(subAccount, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		eToken.mint(subAccount, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol:221]: Is mint amount `uint256(-1)` possible?

File: contracts/mainnet/connectors/euler/main.sol [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+		IEulerDToken dToken = IEulerDToken(markets.underlyingToDToken(_token));
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		_amt = _amt == type(uint256).max
+			? dToken.balanceOf(address(this))
+			: _amt;
+
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		eToken.burn(subAccount, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		eToken.burn(subAccount, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 12]
```diff
@@ -0,0 +1,552 @@
...
+
+		setUint(setId, _amt);
+
+		_eventName = "LogBurn(uint256,address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(subAccount, token, _amt, getId, setId);
+	}
+
+	/**
+	 * @dev ETransfer ETH/ERC20_Token.
+	 * @notice ETransfer deposits from account to another.
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
	 * @notice ETransfer deposits from one sub-account to another.
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+
+		bool isEth = token == ethAddr ? true : false;
+		address _token = isEth ? wethAddr : token;
+
+		IEulerEToken eToken = IEulerEToken(markets.underlyingToEToken(_token));
+
+		_amt = _amt == type(uint256).max
+			? eToken.balanceOf(address(this))
+			: _amt;
+
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: eTokens are interest-bearing tokens, so it increases over time. Better to use `balanceOfUnderlying` if the user wants to transfer the whole eTokens amount, including interest accumulated.

File: contracts/mainnet/connectors/euler/main.sol [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+		if (isEth) convertEthToWeth(isEth, TokenInterface(_token), _amt);
+
+		address _subAccountToAddr = getSubAccount(address(this), subAccountTo);
+
+		eToken.transfer(_subAccountToAddr, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		eToken.transfer(_subAccountToAddr, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 15]
```diff
@@ -0,0 +1,552 @@
...
+			token,
+			_amt,
+			getId,
+			setId
+		);
+	}
+
+	/**
+	 * @dev DTransfer ETH/ERC20_Token.
+	 * @notice DTransfer deposits from account to another.
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
	 * @notice DTransfer deposits from one sub-account to another.
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 16]
```diff
@@ -0,0 +1,552 @@
...
+			_amt,
+			getId,
+			setId
+		);
+	}
+
+	/**
+	 * @dev Approve debt.
+	 * @notice Approves receiver to take debt.
+	 * @param subAccountId Subaccount number
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol:380]:
```suggestion
	 * @param subAccountId Subaccount number that will receive debt.
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+	{
+		uint256 _amt = getUint(getId, amt);
+
+		bool isEth = token == ethAddr;
+		address _token = isEth ? wethAddr : token;
+
+		IEulerDToken dToken = IEulerDToken(markets.underlyingToDToken(_token));
+		_amt = _amt == type(uint256).max
+			? dToken.balanceOf(address(this))
+			: _amt;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: No need for this here, `Euler` supports uint256(-1). 
https://github.com/euler-xyz/euler-contracts/blob/master/contracts/modules/DToken.sol#L179

File: contracts/mainnet/connectors/euler/main.sol [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+
+		dToken.approveDebt(subAccountId, debtReceiver, _amt);
+
+		setUint(setId, _amt);
+
+		_eventName = "LogApproveDebt(uint256,address,address,uint256)";
+		_eventParam = abi.encode(subAccountId, debtReceiver, token, amt);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		_eventParam = abi.encode(subAccountId, debtReceiver, token, _amt);
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 19]
```diff
@@ -0,0 +1,552 @@
...
+
+	struct swapParams {
+		uint256 subAccountFrom;
+		uint256 subAccountTo;
+		address buyAddr;
+		address sellAddr;
+		uint256 sellAmt;
+		uint256 unitAmt;
+		bytes callData;
+	}
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: Can we move these to helpers.sol

File: contracts/mainnet/connectors/euler/main.sol [hunk 20]
```diff
@@ -0,0 +1,552 @@
...
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _length = tokens.length;
+		require(_length > 0, "0-markets-not-allowed");
+
+		for (uint256 i = 0; i < _length; i++) {
+			bool isEth = tokens[i] == ethAddr;
+			address _token = isEth ? wethAddr : tokens[i];
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
			address _token = tokens[i] == ethAddr ? wethAddr : tokens[i];
```

File: contracts/mainnet/connectors/euler/main.sol [hunk 21]
```diff
@@ -0,0 +1,552 @@
...
+
+			IEulerEToken eToken = IEulerEToken(
+				markets.underlyingToEToken(_token)
+			);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: I think this is unused.

File: contracts/mainnet/connectors/euler/main.sol [hunk 22]
```diff
@@ -0,0 +1,552 @@
...
+	 * @param subAccountId Subaccount number
+	 * @param token token address
+	 */
+	function exitMarket(uint256 subAccountId, address token)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		bool isEth = token == ethAddr;
+		address _token = isEth ? wethAddr : token;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]:
```suggestion
		address _token = token == ethAddr ? wethAddr : token;
```

File: contracts/mainnet/connectors/euler/variables.sol [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+pragma solidity ^0.7.0;
+import "./interface.sol";
+
+contract Variables {
+	address internal constant EULER_MAINNET =
+		0x27182842E098f60e3D576794A5bFFb0777E025d3;
+	IEulerMarkets internal constant markets =
+		IEulerMarkets(0x3520d5a913427E6F0D6A83E07ccD4A4da316e4d3);
+	IEulerSwap internal constant swapExec =
+		IEulerSwap(0x7123C8cBBD76c5C7fCC9f7150f23179bec0bA341);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/variables.sol]: Should be declared in `helpers.sol`

File: contracts/mainnet/connectors/euler/main.sol (hunk 6 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Done!

File: contracts/mainnet/connectors/euler/main.sol (hunk 7 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Done!

File: contracts/mainnet/connectors/euler/main.sol (hunk 17 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Nice! Updated.

File: contracts/mainnet/connectors/euler/variables.sol (hunk 23 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/variables.sol]: Done

File: contracts/mainnet/connectors/euler/main.sol (hunk 19 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Yes, done

File: contracts/mainnet/connectors/euler/main.sol (hunk 21 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Oh yes, it is unused. Removed.

File: contracts/mainnet/connectors/euler/main.sol (hunk 10 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol:221]: `mint` is like leverage. So I believe, no. Although there can be a workaround, if it's `uint256(-1)`, then we calculate the safe Max for the position and send that value. I think we can skip it.

Comment (pradyuman-verma) [reply contracts/mainnet/connectors/euler/main.sol:221]: Ohh okay

File: contracts/mainnet/connectors/euler/main.sol (hunk 13 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: `eTokens` are like compound tokens, their value won't change, but the value of the underlying asset would change. So wouldn't transferring the whole `eTokens` amount also mean transferring the interest with it?

File: contracts/mainnet/connectors/euler/variables.sol [hunk 24]
```diff
@@ -0,0 +1,12 @@
+//SPDX-License-Identifier: MIT
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/variables.sol]: We don't need this file right?

File: contracts/mainnet/connectors/euler/interface.sol [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+interface IEulerSwap {
+	struct Swap1InchParams {
+		uint256 subAccountIdIn;
+		uint256 subAccountIdOut;
+		address underlyingIn;
+		address underlyingOut;
+		uint256 amount;
+		uint256 amountOutMinimum;
+		bytes payload;
+	}
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/interface.sol]: We can keep struct out of the interface.

File: contracts/mainnet/connectors/euler/main.sol [hunk 26]
```diff
@@ -0,0 +1,523 @@
...
+
+		bool isEth = token == ethAddr ? true : false;
+
+		address _token = isEth ? wethAddr : token;
+		IEulerDToken borrowedDToken = IEulerDToken(
+			markets.underlyingToDToken(_token)
+		);
+
+		if (isEth) {
+			_amt = _amt == uint256(-1) ? borrowedDToken.balanceOf(address(this)) : _amt;
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/euler/main.sol]: This should be in line 180. 
```suggestion
_amt = _amt == uint256(-1) ? borrowedDToken.balanceOf(address(this)) : _amt;
if (isEth) {
```

File: contracts/mainnet/connectors/euler/main.sol (hunk 13 above)
Comment (pradyuman-verma) [reply contracts/mainnet/connectors/euler/main.sol]: https://github.com/euler-xyz/euler-contracts/blob/master/contracts/modules/EToken.sol#L79

Comment (pradyuman-verma) [reply contracts/mainnet/connectors/euler/main.sol]: `balanceOfUnderlying` returns the balance in token decimals, while `balanceOf` returns it in its own internal unit.

Comment (pradyuman-verma) [reply contracts/mainnet/connectors/euler/main.sol]: I think we should confirm this difference with the Euler team?

Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/main.sol]: Resolved!

File: contracts/mainnet/connectors/euler/variables.sol (hunk 24 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/variables.sol]: Oh yes! removed

File: contracts/mainnet/connectors/euler/interface.sol (hunk 25 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/euler/interface.sol]: Done

**Review by pradyuman-verma — APPROVED** [review]

**Comment by shriyatyagii** [comment]
Euler Connector :

- Address - [0xb12bA67bac0A287d8C0634f2C6e5CEd833e1896d](https://etherscan.io/address/0xb12bA67bac0A287d8C0634f2C6e5CEd833e1896d#code)
- Name - **EULER-A**

**Pull request merged**